scipy/stats/stats.py

```diff
@@ -5717,33 +5770,167 @@ def ttest_ind(a, b, axis=0, equal_var=True, nan_policy='propagate',
-                             "supported by one-sided alternatives.")
+                             "supported by permutation tests or one-sided"
+                             "asymptotic tests.")
         a = ma.masked_invalid(a)
         b = ma.masked_invalid(b)
         return mstats_basic.ttest_ind(a, b, axis, equal_var)
 
     if a.size == 0 or b.size == 0:
         return _ttest_nans(a, b, axis, Ttest_indResult)
 
-    v1 = np.var(a, axis, ddof=1)
-    v2 = np.var(b, axis, ddof=1)
-    n1 = a.shape[axis]
-    n2 = b.shape[axis]
+    if permutations:
+        if int(permutations) != permutations or permutations < 0:
+            raise ValueError("Permutations must be a positive integer.")
 
-    if equal_var:
-        df, denom = _equal_var_ttest_denom(v1, n1, v2, n2)
-    else:
-        df, denom = _unequal_var_ttest_denom(v1, n1, v2, n2)
+        res = _permutation_ttest(a, b, permutations=permutations,
+                                 axis=axis, equal_var=equal_var,
+                                 nan_policy=nan_policy,
+                                 random_state=random_state,
+                                 alternative=alternative)
 
-    res = _ttest_ind_from_stats(np.mean(a, axis), np.mean(b, axis), denom, df,
-                                alternative)
+    else:
+        v1 = np.var(a, axis, ddof=1)
+        v2 = np.var(b, axis, ddof=1)
+        n1 = a.shape[axis]
+        n2 = b.shape[axis]
 
+        if equal_var:
+            df, denom = _equal_var_ttest_denom(v1, n1, v2, n2)
+        else:
+            df, denom = _unequal_var_ttest_denom(v1, n1, v2, n2)
+        res = _ttest_ind_from_stats(np.mean(a, axis), np.mean(b, axis),
+                                    denom, df, alternative)
     return Ttest_indResult(*res)
 
 
+def _broadcast_concatenate(xs, axis):
+    """Concatenate arrays along an axis with broadcasting"""
+    # move the axis we're concatenating along to the end
+    xs = [np.swapaxes(x, axis, -1) for x in xs]
+    # determine final shape of all but the last axis
+    shape = np.broadcast(*[x[..., 0] for x in xs]).shape
+    # broadcast along all but the last axis
+    xs = [np.broadcast_to(x, shape + (x.shape[-1],)) for x in xs]
+    # concatenate along last axis
+    res = np.concatenate(xs, axis = -1)
+    # move the last axis back to where it was
+    res = np.swapaxes(res, axis, -1)
+    return res
+
+
+def _data_permutations(data, n, axis=-1, random_state=None):
+    """Vectorized permutation of data"""
+    random_state = check_random_state(random_state)
+    if axis < 0:  # we'll be adding a new dimension at the end
+        axis = data.ndim + axis
+
+    # prepare permutation indices
+    m = data.shape[axis]
+    n_max = float_factorial(m)  # number of distinct permutations
+
+    if n < n_max:
+        indices = np.array([random_state.permutation(m) for i in range(n)]).T
+    else:
+        n = n_max
+        indices = np.array(list(permutations(range(m)))).T
```

Comment:
I think it's OK to merge as is, but afterwards (at some point) I will submit a PR that makes this much more efficient. We don't need all of these permutations because the t-statistic doesn't care about the order of the numbers within the groups, just which group the numbers are in. Since I added the exact test while thinking about the permutation test, I didn't think that through. Anyway, this should give the same result; it's just that taking combinations is way more efficient.